flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/AppendOnlyCompactManager.java

```diff
@@ -0,0 +1,128 @@
+
+    public AppendOnlyCompactManager(
+            ExecutorService executor,
+            LinkedList<DataFileMeta> toCompact,
+            int minFileNum,
+            int maxFileNum,
+            long targetFileSize,
+            CompactRewriter rewriter) {
+        super(executor);
+        this.toCompact = toCompact;
+        this.maxFileNum = maxFileNum;
+        this.minFileNum = minFileNum;
+        this.targetFileSize = targetFileSize;
+        this.rewriter = rewriter;
+    }
+
+    @Override
+    public void submitCompaction() {
+        if (taskFuture != null) {
+            throw new IllegalStateException(
+                    "Please finish the previous compaction before submitting new one.");
+        }
+        pickCompactBefore()
+                .ifPresent(
+                        (compactBefore) ->
+                                taskFuture =
+                                        executor.submit(
+                                                new CompactTask() {
+                                                    @Override
+                                                    protected CompactResult compact()
+                                                            throws Exception {
+                                                        collectBeforeStats(compactBefore);
+                                                        List<DataFileMeta> compactAfter =
+                                                                rewriter.rewrite(compactBefore);
+                                                        collectAfterStats(compactAfter);
+                                                        return result(compactBefore, compactAfter);
+                                                    }
+                                                }));
+    }
+
+    @VisibleForTesting
+    Optional<List<DataFileMeta>> pickCompactBefore() {
+        long totalFileSize = 0L;
+        int fileNum = 0;
+        int releaseCtr = 0;
+        for (int i = 0; i < toCompact.size(); i++) {
+            DataFileMeta file = toCompact.get(i);
+            totalFileSize += file.fileSize();
+            fileNum++;
+            int pos = i - fileNum + 1;
+            if ((totalFileSize >= targetFileSize && fileNum >= minFileNum)
+                    || fileNum >= maxFileNum) {
+                // trigger compaction for [pos, i]
+                List<DataFileMeta> compactBefore = new ArrayList<>(toCompact.subList(pos, i + 1));
+                // files in [0, pos - 1] can be released immediately
+                // [pos, i] should be released after compaction finished
+                for (int j = 0; j <= pos - 1; j++) {
+                    toCompact.pollFirst();
+                }
+                return Optional.of(compactBefore);
+            } else if (totalFileSize >= targetFileSize) {
+                // this is equivalent to shift one pos to right
+                fileNum--;
```

Comment:
> Why not poll first directly?

Because we still need to iterate on `toCompact`. I think we can poll `toCompact` only when we find a qualified sub-list to be compacted.